### ISSUE TITLE
End Session Callback - A second operation started on this context before a previous asynchronous operation completed

### ISSUE THREAD

**Issue by AnEmptyReece** [IdentityServer/IdentityServer3#2253]
The following exception occurs when hitting the EndSession callback as part of a Logout (since the new Http Logout spec was introduced in v2.2)

```
IdentityServer3.Core.Endpoints.EndSessionController: 2015-12-02 16:01:31,272 [10] INFO  - End session callback requested
IdentityServer3.Core.Configuration.Hosting.LogProviderExceptionLogger: 2015-12-02 16:01:31,300 [8] ERROR - Unhandled exception
System.NotSupportedException: A second operation started on this context before a previous asynchronous operation completed. Use 'await' to ensure that any asynchronous operations have completed before calling another method on this context. Any instance members are not guaranteed to be thread safe.
   at System.Data.Entity.Internal.ThrowingMonitor.EnsureNotEntered()
   at System.Data.Entity.Core.Objects.ObjectQuery`1.System.Data.Entity.Infrastructure.IDbAsyncEnumerable<T>.GetAsyncEnumerator()
   at System.Data.Entity.Infrastructure.IDbAsyncEnumerableExtensions.<SingleOrDefaultAsync>d__38`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at IdentityServer3.EntityFramework.ClientStore.<FindClientByIdAsync>d__2.MoveNext() in c:\ballen\github\identity\IdSrv3\IdentityServer3.EntityFramework\Source\Core.EntityFramework\Stores\ClientStore.cs:line 39
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IdentityServer3.Core.Endpoints.EndSessionController.<GetClientEndSessionUrlsAsync>d__14.MoveNext() in c:\local\identity\server3\Core\source\Core\Endpoints\Connect\EndSessionController.cs:line 183
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at IdentityServer3.Core.Endpoints.EndSessionController.<LogoutCallback>d__a.MoveNext() in c:\local\identity\server3\Core\source\Core\Endpoints\Connect\EndSessionController.cs:line 131
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.System.Web.Http868683.TaskHelpersExtensions.<CastToObject>d__3`1.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__0.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__0.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.AuthenticationFilterResult.<ExecuteAsync>d__0.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext() in c:\local\identity\server3\Core\source\Core\Services\Default\DefaultClaimsProvider.cs:line 0
Exception thrown: 'System.Data.Entity.Validation.DbEntityValidationException' in EntityFramework.dll
Exception thrown: 'System.Data.Entity.Validation.DbEntityValidationException' in mscorlib.dll
```

I've tracked this down to being caused by the following section of code in the EndSessionController:

```
        private async Task<IEnumerable<string>> GetClientEndSessionUrlsAsync()
        {
            // read client list to get URLs for client logout endpoints
            var clientIds = _clientListCookie.GetClients();

            // fetch the clients in parallel
            var tasks =
                from clientId in clientIds
                select _clientStore.FindClientByIdAsync(clientId);
            await Task.WhenAll(tasks);
```

This seems to be because running the tasks in parallel means that multiple operations will occur on the same DbContext (as the exception suggests)

A quick change to the code to something like this:

```
            var clients = new List<Client>();
            foreach (var clientId in clientIds)
            {
                clients.Add(await this._clientStore.FindClientByIdAsync(clientId));
            }
```

Does seem to resolve the issue.

Ideally, perhaps a new IClientStore method would be created to Get a list of clients for a supplied list of Ids - that way only one database call would be required anyway.

I'll look at the best approach and hopefully submit a PR over the next week or so.

**Comment by brockallen** [comment]
I don't know if MARS in the connection string would help or not (with thw original LINQ query).

**Comment by AnEmptyReece** [comment]
Just tried it and it makes no difference unfortunately.

**Comment by brockallen** [comment]
Ok, so sounds like the use of the parallel query was a bad idea. PR away, when you have time please!

**Comment by AnEmptyReece** [comment]
#2254 PR created. It's my first one so hopefully everything is in order!

**Comment by brockallen** [comment]
Merged on dev. thx.

**Comment by snothub** [comment]
Guys, after this upgrade to 2.2.2 (from 2.1.1) I'm now getting errors on userinfo endpoint from TokenValidator class all over. 

"Checking for expected scope openid failed"
  '\"ExpectedScope\": \"openid\",
  \"Claims\": {\r\n    \
  	"iss\": ...
  	"aud\": ...
  	"exp\": ...
  	"nbf\": ...
  	"client_id\": \"ro-client\",\r\n    \
  	"scope\": [\r\n      \"\\\"offline_access\\\"\",\r\n      \"\\\"openid\\\"\",\r\n      \"\\\"profile\\\"\"\r\n    ],\r\n    
etc


Downgrading back to 2.1.1 works again.

**Comment by leastprivilege** [comment]
i need more info to repro

**Comment by leastprivilege** [comment]
..and probably also a new issue.

**Comment by leastprivilege** [comment]
ok - found it. investigating

**Comment by leastprivilege** [comment]
https://github.com/IdentityServer/IdentityServer3/issues/2261
